### PR TITLE
Activate pkg envs from JLSOFiles

### PR DESCRIPTION
**Pull request by rofinn**
- Closes #28 
- More graceful failures on failure during _upgrade_env (needed when tagging new releases)

### CODE REVIEW

File: src/metadata.jl [hunk 1]
```diff
@@ -52,3 +52,20 @@ function _image()
...
 end
+
+function Pkg.activate(jlso::JLSOFile, path; kwargs...)
+    ispath(path) || mkpath(path)
+    open(io -> Pkg.TOML.print(io, jlso.project), joinpath(path, "Project.toml"), "w")
+    open(io -> Pkg.TOML.print(io, jlso.manifest), joinpath(path, "Manifest.toml"), "w")
+    Pkg.activate(path; kwargs...)
+end
+
+function Pkg.activate(f::Function, jlso::JLSOFile, path; kwargs...)
```
Comment (oxinabox) [inline src/metadata.jl]: should make `path` optional and defaulting to `mktmp()`
here and above

**Comment by oxinabox** [comment]
Does this work fine for if something needs to be installed?
Also for if the Manifest points to a custom branch.

File: src/deprecated.jl [hunk 2]
```diff
@@ -97,9 +97,21 @@ function _upgrade_env(pkgs::Dict)
...
+                    Pkg.PackageSpec(; name=key, version=value) for (key, value) in pkgs
+                ])
+            catch e
+                # Warn about failure and fallback to simply trying to install the pacakges
+                # without version constraints.
+                warn(LOGGER) do
+                    "Failed to construct an environment with the provide package version " *
+                    "($pkgs): $e.\n Falling back to simply adding the packages."
+                end
+                Pkg.add([Pkg.PackageSpec(; name=key) for (key, value) in pkgs])
```
Comment (nickrobinson251) [inline src/deprecated.jl]: What behaviour does this have for packages with the same name (distinguishable by UUID)?

Comment (rofinn) [reply src/deprecated.jl]: This PR won't change the behaviour, but `Pkg.add` will prompt the user to select the registry where the package lives.

File: src/metadata.jl (hunk 1 above)
Comment (rofinn) [reply src/metadata.jl]: 👌

File: src/deprecated.jl (hunk 2 above)
Comment (oxinabox) [reply src/deprecated.jl]: I guess because this is for handling old files, and thus we don't have the UUID recorded, that is the best we can do

Comment (rofinn) [reply src/deprecated.jl]: That's correct, there's only so much we can do when upgrading from older JLSO format versions.

File: src/metadata.jl [hunk 3]
```diff
@@ -52,3 +52,20 @@ function _image()
 
     return _CACHE[:IMAGE]
 end
+
+function Pkg.activate(jlso::JLSOFile, path=tempdir(); kwargs...)
+    ispath(path) || mkpath(path)
```
Comment (oxinabox) [inline src/metadata.jl]: The fact that we are doing this and not:
```suggestion
    mkdir(p"a"; recursive=true, exist_ok=true)
```
suggests something must be wrong in the FilePaths API.

**Comment by oxinabox** [comment]
None of the tests are integration tests that actually check can load a file with that enviroment.
Is it worth writing them or is it enough to be sure we are getting the Project.toml right?

I think it is probably enough,
as otherwise we need to run things in seperate processes since we have already loaded some of the modules

File: src/metadata.jl (hunk 3 above)
Comment (rofinn) [reply src/metadata.jl]: We don't depend on FilePathsBase here which I think is a reasonable choice for now.

**Comment by oxinabox** [comment]
I mean we don't test actually activating an old enviroment from JLSO then opening the file.
But I think that is ok

File: src/metadata.jl (hunk 3 above)
Comment (oxinabox) [reply src/metadata.jl]: it does indeed not
```
shell> mkdir foo

shell> chmod 660 foo

shell> ls -lah
...
drw-rw----   2 oxinabox  staff    64B  5 Nov 18:24 foo

julia> mkpath("foo")
"foo"

shell> ls -lah
...
drw-rw----   2 oxinabox  staff    64B  5 Nov 18:24 foo

help?> mkpath
search: mkpath

  mkpath(path::AbstractString; mode::Unsigned = 0o777)

  Create all directories in the given path, with permissions mode. mode defaults to 0o777, modified by the current file creation mask. Return path.
```

I

Comment (rofinn) [reply src/metadata.jl]: Looks like the answer that question is no https://github.com/JuliaLang/julia/blob/02deed30bfda45250ab0b6615b780e205e4811d9/base/file.jl#L221

**Pull request merged**